docs/book/src/developer/repository-layout.md

```diff
@@ -1,3 +1,175 @@
+
+### Tools
+
+[~/hack](https://github.com/kubernetes-sigs/cluster-api/tree/main/hack)
+
+This folder has scripts used for building, testing and developer workflow. Its subfolder [~/hack/observability](https://github.com/kubernetes-sigs/cluster-api/tree/main/hack/observability) is a place for resources related to  observability like Loki, graphing,  prometheus and promtail.
+
+[~/scripts](https://github.com/kubernetes-sigs/cluster-api/tree/main/scripts)
+
+This folder consists of CI scripts related to setup, build and e2e tests. These are mostly called by CI jobs.
+
+### CAPI as a library
+
+[~/util](https://github.com/kubernetes-sigs/cluster-api/tree/main/util)
+
+This folder contains utilities which are meant to be  used in multiple CAPI modules, providers and users of CAPI.
+
+[~/feature](https://github.com/kubernetes-sigs/cluster-api/tree/main/feature)
+
+This package provides feature gate management used in Cluster API as well as providers. This implementation of feature gates is shared across all providers.
+
+[~/errors](https://github.com/kubernetes-sigs/cluster-api/tree/main/errors)
+
+This is a place for defining errors returned by CAPI. Error types defined here can be used by users of CAPI and the providers.
+
+### Experimental features
+
+[~/exp](https://github.com/kubernetes-sigs/cluster-api/tree/main/exp)
+
+This folder contains experimental features of CAPI. Experimental features are unreliable until they are promoted to the main repository. Each experimental feature is supposed to be present in a subfolder of [~/exp](https://github.com/kubernetes-sigs/cluster-api/tree/main/exp) folder. For e.g. ClusterResourceSet is present inside [~/exp/addons](https://github.com/kubernetes-sigs/cluster-api/tree/main/exp/addons) folder. Historically, machine pool resources are not present in a sub-directory. Migrating them to a subfolder like `~/exp/machinepools` is still pending as it can potentially break existing users who are relying on existing folder structure.
+
+These experimental features are kubebuilder projects having standard folder structure except that the CRDs for these experimental features are present in [config](#configuration) folder of the main repository. These CRDs are deployed along with CAPI cluster. However, these features can be enabled and disabled using feature gates. 
+
+### Webhooks
+
+In the [api](#api) folder, we have simple webhook implementation consisting of validators and defaults for various types. Those implementations do not serve custom use cases for e.g. we do not have a client to the API Server to read objects and perform certain validations. 
+
+[~/internal/webhooks](https://github.com/kubernetes-sigs/cluster-api/tree/main/internal/webhooks)
+
+This directory contains custom webhook interface which includes client as well. This allows us to perform complex validations during CRUD operations by reading different types from API Server. 
```

Comment:
```suggestion
This directory contains the implementation of some of the Cluster API webhooks. The internal implementation means that the methods supplied by this package can not be imported by external code bases.
```

Reply:
We have two interfaces for webhooks
simple webhooks present in ~/api folder (without client)
& custom webhooks present in ~/webhooks folder (with client). Depending on complexity of use case, any new webhook would go in one of these folders.
This information might be useful for contributors, so can we include it?

Reply:
The guidance on where a webhook implementation ends up isn't that clearly defined IMO, though it does directly line up with what we did with the Cluster and ClusterClass webhooks. I'd prefer to keep information that can change easily out of this doc so it's easier to maintain.